src/endpoint/rules.ts

```diff
@@ -1,174 +1,168 @@
-	 */
-	id?: string
-	left: Operand
-	aggregation?: ConditionAggregationMode
-}
-
-export interface TimedSimpleCondition {
-	id?: string
-	left: Operand
-	right: Operand
-	interval: Interval
-	aggregation?: ConditionAggregationMode
+/**
+ * A condition that returns true when its evaluation resolves to true and the previous evaluation
+ * resolved to false.
+ */
+export interface ChangesCondition extends BasicCondition {
+	id: string
+	operand?: RuleOperand
 }
 
-export enum ArmState {
-	ArmedStay = 'ArmedStay',
-	ArmedAway = 'ArmedAway',
-	Disarmed = 'Disarmed',
+/**
+ * A condition that returns true if its evaluation would return true within the specified duration.
+ */
+export interface WasCondition extends BasicCondition {
+	id: string
+	operand?: RuleOperand
+	duration: RuleInterval
 }
 
-export interface SecurityState {
-	armState: ArmState
-	monitoring?: boolean
+/**
+ * A condition that returns true if its evaluation is true after the specified duration.
+ */
+export interface RemainsCondition extends BasicCondition {
+	id: string
+	duration: RuleInterval
+	operand?: RuleOperand
 }
 
-export interface Condition {
-	and?: Condition[]
-	or?: Condition[]
-	not?: Condition
+export interface BasicCondition {
+	and?: RuleCondition[]
+	or?: RuleCondition[]
+	not?: RuleCondition
 	equals?: SimpleCondition
 	greaterThan?: SimpleCondition
 	greaterThanOrEquals?: SimpleCondition
 	lessThan?: SimpleCondition
 	lessThanOrEquals?: SimpleCondition
 	between?: BetweenCondition
-	dropsBelow?: SimpleCondition
-	dropsToOrBelow?: SimpleCondition
-	risesAbove?: SimpleCondition
-	risesToOrAbove?: SimpleCondition
 }
 
-export enum SubscriptionMode {
-	Auto = 'Auto',
-	Always = 'Always',
-	Never = 'Never'
-}
-
-export enum TimeReference {
-	Now = 'Now',
-	Midnight = 'Midnight',
-	Sunrise = 'Sunrise',
-	Noon = 'Noon',
-	Sunset = 'Sunset'
-}
-
-export interface TimeOperand {
-	/**
-	 * A java time zone ID reference
-	 */
-	timeZoneId?: string
-	reference: TimeReference
-	offset?: Interval
+export interface RuleCondition {
+	and?: RuleCondition[]
+	or?: RuleCondition[]
+	not?: RuleCondition
+	equals?: SimpleCondition
+	greaterThan?: SimpleCondition
+	greaterThanOrEquals?: SimpleCondition
+	lessThan?: SimpleCondition
+	lessThanOrEquals?: SimpleCondition
+	between?: BetweenCondition
+	changes?: ChangesCondition
+	remains?: RemainsCondition
+	was?: WasCondition
 }
 
-export enum OperandAggregationMode {
-	None = 'None',
-	Avg = 'Avg',
-	Sum = 'Sum',
-	Min = 'Min',
-	Max = 'Max',
-	Least = 'Least'
-}
+export type TimeReference ='Now' | 'Midnight' | 'Sunrise' | 'Noon' | 'Sunset'
+export type DateReference = 'Today'
+export type OperandAggregationMode = 'None'
+export type TriggerMode = 'Auto' | 'Always' | 'Never'
 
 export interface ArrayOperand {
-	operands: Operand[]
+	operands: RuleOperand[]
 	aggregation?: OperandAggregationMode
 }
 
-export class MapOperand extends null<string, Operand> {
-
+export interface MapOperand {
+	[name: string]: RuleOperand
```

Comment:
To be safer, I would make this `[name: string]: RuleOperand | undefined` since Typescript will assume all possible strings map to a value.

Reply:
Sorry, I don't understand why this would make a difference. Do you have a link or something that might help me?

Reply:
https://github.com/microsoft/TypeScript/issues/13778